pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1103,46 +835,35 @@ protected IdealState ensureAllPartitionsConsuming(final TableConfig tableConfig,
     //    a. Create a new segment (with the next seq number)
     //       and restart consumption from the same offset (if possible) or a newer offset (if realtime stream does not have the same offset).
     //       In latter case, report data loss.
-    for (Map.Entry<Integer, LLCRealtimeSegmentZKMetadata[]> entry : partitionToLatestMetadata.entrySet()) {
-      int partition = entry.getKey();
-      LLCRealtimeSegmentZKMetadata[] latestMetadataArray = entry.getValue();
-      LLCRealtimeSegmentZKMetadata latestMetadata = latestMetadataArray[0];
-      final String segmentId = latestMetadata.getSegmentName();
-      final LLCSegmentName segmentName = new LLCSegmentName(segmentId);
-
-      Map<String, Map<String, String>> mapFields = idealState.getRecord().getMapFields();
-      if (mapFields.containsKey(segmentId)) {
+    for (Map.Entry<Integer, LLCRealtimeSegmentZKMetadata> entry : latestSegmentZKMetadataMap.entrySet()) {
+      int partitionId = entry.getKey();
+      LLCRealtimeSegmentZKMetadata latestSegmentZKMetadata = entry.getValue();
+      String latestSegmentName = latestSegmentZKMetadata.getSegmentName();
+      LLCSegmentName latestLLCSegmentName = new LLCSegmentName(latestSegmentName);
+
+      Map<String, String> instanceStateMap = instanceStatesMap.get(latestSegmentName);
+      if (instanceStateMap != null) {
         // Latest segment of metadata is in idealstate.
-        Map<String, String> instanceStateMap = idealState.getInstanceStateMap(segmentId);
-        if (instanceStateMap.values().contains(PinotHelixSegmentOnlineOfflineStateModelGenerator.CONSUMING_STATE)) {
-          if (latestMetadata.getStatus().equals(CommonConstants.Segment.Realtime.Status.DONE)) {
+        if (instanceStateMap.values().contains(RealtimeSegmentOnlineOfflineStateModel.CONSUMING)) {
+          if (latestSegmentZKMetadata.getStatus() == Status.DONE) {
 
             // step-1 of commmitSegmentMetadata is done (i.e. marking old segment as DONE)
             // but step-2 is not done (i.e. adding new metadata for the next segment)
             // and ideal state update (i.e. marking old segment as ONLINE and new segment as CONSUMING) is not done either.
-            if (isTooSoonToCorrect(tableNameWithType, segmentId, now)) {
+            if (!isExceededMaxSegmentCompletionTime(realtimeTableName, latestSegmentName, currentTimeMs)) {
               continue;
             }
-            LOGGER.info("{}:Repairing segment for partition {}. "
-                    + "Old segment metadata {} has status DONE, but segments are still in CONSUMING state in ideal STATE",
-                tableNameWithType, partition, segmentId);
-
-            LLCSegmentName newLLCSegmentName = makeNextLLCSegmentName(segmentName, partition, now);
-            LOGGER.info("{}: Creating new segment metadata for {}", tableNameWithType,
```

Comment:
Can we add this log back in some form? Basically to know which new segment metadata were created by the ValidationManager

Reply:
This is logged in `createNewSegmentZKMetadata()`